board.cpp

```diff
@@ -53,55 +53,153 @@ bool Board::isMovable(const QModelIndex &index)
+                 next += m_directions[DOWN])
+            {
+                auto prev = next + m_directions[UP];
+                std::swap(m_data[next.x()][next.y()], m_data[prev.x()][prev.y()]);
+            }
+        }
+    }
+}
+
+void Board::fillEmpty()
+{
+    auto isValid = [this] (QPoint& p)
+    {
+        return ( (p.x() < 0)
+                 || (p.y() < 0)
+                 || (p.x() > m_dimension - 1)
+                 || (p.y() > m_dimension - 1) ? false : true );
+    };
+
+    QPoint current;
+    QVector<QPoint> toFill;
+    QVector<QVector<bool>> visited(m_dimension, QVector<bool>(m_dimension, false));
     QQueue<QPoint> queue;
 
-    visited[inx(current)] = true;
+    bool found = false;
+    for (int i = 0; !found && i < m_data.size(); ++i)
+    {
+        for (int j = 0; j < m_data.first().size(); ++j)
+        {
+            if (m_data[i][j].color() == Qt::transparent)
+            {
+                found = true;
+                current = {i, j};
+                break;
+            }
+        }
+    }
+
+    visited[current.x()][current.y()] = true;
     queue.enqueue(current);
-    forPopping.append(current);
+    toFill.append(current);
 
     while (!queue.empty())
     {
-        current = queue.dequeue();
-        qDebug() << current.x() << " " << current.y() << "\n";
+        current = queue.head();
         queue.dequeue();
 
-        for (const auto& direction : directions)
+        for (const auto& direction : m_directions)
         {
-            auto forCheck = current + direction;
-            if(isValid(forCheck) && !visited[inx(forCheck)] &&  m_data[current.x()][current.y()] == m_data[forCheck.x()][forCheck.y()])
+            auto check = current + direction;
+            if (	isValid(check)
+                    && !visited[check.x()][check.y()]
+                    && m_data[check.x()][check.y()].color() == Qt::transparent)
             {
-                visited[inx(forCheck)] = true;
-                forPopping.append(forCheck);
-                queue.enqueue(forCheck);
+                visited[check.x()][check.y()] = true;
+                toFill.append(check);
+                queue.enqueue(check);
             }
         }
     }
 
-    beginResetModel();
-    for (auto& ball : forPopping)
+    std::random_device rd;
+    std::default_random_engine e(rd());
+
+    Colors availableColors;
+    QColor prevTop;
+    QColor prevLeft;
+
+    for (const auto& cur : toFill)
     {
-        m_data[ball.x()][ball.y()] = Qt::transparent;
+        availableColors = m_colors;
+        prevTop = (cur.x() != 0) ? m_data[cur.x()-1][cur.y()].color() : QColor();
+        prevLeft = (cur.y() != 0) ? m_data[cur.x()][cur.y()-1].color() : QColor();
+        availableColors.removeOne(prevTop);
+        availableColors.removeOne(prevLeft);
+        std::uniform_int_distribution<int> uniform_dist(0, availableColors.size() - 1);
+        m_data[cur.x()][cur.y()] = availableColors[uniform_dist(e)];
```

Comment:
This code is duplicated below. Only cycle differs.
Let's think about moving it into separate method so we can remove duplicated code